vmdb/app/controllers/api_controller/providers.rb

```diff
@@ -0,0 +1,126 @@
+    end
+
+    def refresh_resource_providers(type, id = nil, _data = nil)
+      raise BadRequestError, "Must specify an id for refreshing a #{type} resource" unless id
+
+      api_action(type, id) do |klass|
+        provider = resource_search(id, type, klass)
+        api_log_info("Refreshing #{provider_ident(provider)}")
+
+        refresh_provider(provider)
+      end
+    end
+
+    def delete_resource_providers(type, id = nil, _data = nil)
+      raise BadRequestError, "Must specify an id for deleting a #{type} resource" unless id
+
+      api_action(type, id) do |klass|
+        provider = resource_search(id, type, klass)
+        api_log_info("Deleting #{provider_ident(provider)}")
+
+        destroy_provider(provider)
+      end
+    end
+
+    private
+
+    def provider_ident(provider)
+      "Provider id:#{provider.id} name:'#{provider.name}'"
+    end
+
+    def fetch_provider_klass(klass, data)
+      supported_types = klass.supported_subclasses.collect(&:name)
+      types_string    = supported_types.join(", ")
+
+      unless data.key?("type")
+        raise BadRequestError, "Must specify a provider type, supported types are: #{types_string}"
+      end
+      type = data["type"]
+      unless supported_types.include?(type)
+        raise BadRequestError, "Invalid provider type #{type} specified, supported types are: #{types_string}"
+      end
+      klass.supported_subclasses.detect { |p| p.name == data["type"] }
+    end
+
+    def create_provider(data)
+      klass = collection_config[:providers][:klass].constantize
+      create_data = fetch_provider_data(data, :include_zone => true)
+      provider = fetch_provider_klass(klass, data).create!(create_data)
+      update_provider_authentication(provider, data)
+      provider
+    rescue => err
+      provider.destroy if provider
+      raise BadRequestError, "Could not create the new provider - #{err}"
+    end
+
+    def edit_provider(provider, data)
+      update_data = fetch_provider_data(data)
+      provider.update_attributes(update_data) if update_data.present?
+      update_provider_authentication(provider, data)
+      provider
+    rescue => err
+      raise BadRequestError, "Could not update the provider - #{err}"
+    end
+
+    def refresh_provider(provider)
+      desc = "#{provider_ident(provider)} refreshing"
+      provider.refresh_ems
+      action_result(true, desc)
+    rescue => err
+      action_result(false, err.to_s)
+    end
+
+    def destroy_provider(provider)
+      desc = "#{provider_ident(provider)} deleting"
+      task_id = queue_object_action(provider, desc, :method_name => "destroy")
+      action_result(true, desc, :task_id => task_id)
+    rescue => err
+      action_result(false, err.to_s)
+    end
+
+    def update_provider_authentication(provider, data)
+      credentials = data.slice(*CREDENTIAL_ATTRS)
+      provider.update_authentication(:default => credentials.symbolize_keys!) if credentials.present?
```

Comment:
I think this is wrong, but I'm not sure how to resolve.  Not all authentications are the default authentication.  Some providers need multiple sets of authentications.  See the UI code for examples.